Rename AutoNonVariableTypeMode to AutoDispatchBelowAutograd and add a warning.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56424 AutoDispatchBelowAutograd takes no arguments.
* #56423 s/AutoNonVariableTypeMode/AutoDispatchBelowAutograd/
* **#56422 Rename AutoNonVariableTypeMode to AutoDispatchBelowAutograd and add a warning.**
* #56434 Switch assertWarnsOnceRegex logic to check any instead of all.

Differential Revision: [D27866608](https://our.internmc.facebook.com/intern/diff/D27866608)

Commit:
Rename AutoNonVariableTypeMode to AutoDispatchBelowAutograd and add a warning.

[ghstack-poisoned]

albanD reviewed (approved):
LGTM

bhosmer commented on aten/src/ATen/core/LegacyTypeDispatch.h:
```diff
@@ -17,7 +17,7 @@ namespace at {
 // NOTE [ Treating Variables as non-Variables in type dispatch ]
 //
-// What exactly does AutoNonVariableType do?  The short answer is, it causes
+// What exactly does AutoDispatchBelowAutograd do?  The short answer is, it causes
```
We don't really use "below" as the name for this concept anywhere else, do we? Not that we should sweat this name too much, but we do use "after" in e.g. `DispatchKeySet::FullAfter`, so maybe `AutoDispatchAfterAutograd`?

ailzhang replied:
@bhosmer I named the other RAII guard `AutoDispatchBelowInplaceOrView` so we should either keep or change them both together I think ;)

bhosmer replied:
Ah you're absolutely right (and my bad for not raising the "below" thing there!) Anyway it's a very minor difference and the note explains exactly what the guard does in any case. Definitely not worth digging up the sidewalk.

bhosmer reviewed (approved):
(sorry, meant to approve - your call on "below" vs "after" naming suggestion ofc 😁 )

Commit:
Update on "Rename AutoNonVariableTypeMode to AutoDispatchBelowAutograd and add a warning."


Differential Revision: [D27866608](https://our.internmc.facebook.com/intern/diff/D27866608)

[ghstack-poisoned]

Commit:
Update on "Rename AutoNonVariableTypeMode to AutoDispatchBelowAutograd and add a warning."


Differential Revision: [D27866608](https://our.internmc.facebook.com/intern/diff/D27866608)

[ghstack-poisoned]
